Add spent column to mobile view

Crazypkr1099 opened the pull request.

Kidglove57 commented:
I opened my own file in the preview. The three columns display perfectly on my iPhone 13 Pro using GBP. No truncating of category names or numbers.

just a thought  - may give a problem with those currencies where even small denominations run to many numbers?

shall0pass commented:
> just a thought - may give a problem with those currencies where even small denominations run to many numbers?

That's a good point. I put a transaction of 2,500,000 in to see what the display looked like.  The numbers do start to run together, though I can't really claim the latest release is too much better in this regard.

Demo.actualbudget.org
![Screenshot_20230904-060842-796](https://github.com/actualbudget/actual/assets/20625555/df269e36-e115-49d0-971d-fa41dc9d1a6b)
![Screenshot_20230904-060727-410](https://github.com/actualbudget/actual/assets/20625555/3d21edbe-003f-4263-9f40-83c16b2227b1)

Kidglove57 commented:
But then I suspect that the currencies we are referring too give the same problem on mobile in any finance app? And even in the current two column mode?

Personally I would love to see the spent column added. Just need the check on a smaller mobile such as the iPhone Mini?

Crazypkr1099 commented:
![ab](https://github.com/actualbudget/actual/assets/11566471/a6fc7544-6802-40f0-aa31-0c53a3098595)

This is what it looks like on iphone mini dimensions

Crazypkr1099 commented:
Recent Commit added flex-basis of 80px to prevent issues with "Projected Savings" as shown in the below photo:
![ab](https://github.com/actualbudget/actual/assets/11566471/f4d5f4c1-2d49-4dce-bde8-8ed2ce20666e)

MatissJanis commented:
👋 Would you mind updating the VRT screenshots in order to get the VRT CI job passing? Instructions can be found here: https://github.com/actualbudget/actual/blob/master/packages/desktop-client/README.md

TLDR:
```
E2E_START_URL=https://192.168.0.178:3001 yarn vrt --update-snapshots
```

MatissJanis commented:
Also: this is iphone SE. Looks pretty broken due to the small screen size. Would it be possible to hide the extra column for such small screens?

<img width="735" alt="Screenshot 2023-09-05 at 21 30 29" src="https://github.com/actualbudget/actual/assets/886567/c65bef22-2b99-43e0-9bbe-50a0422d41de">

Crazypkr1099 commented on packages/desktop-client/src/components/budget/MobileBudgetTableSmall.js:
```diff
@@ -0,0 +1,1092 @@
+import React, { Component, memo, PureComponent, useState } from 'react';
```
@matiss.ja I saw you comment and I totally agree that doing it that was is awful. The problem is that if I want to make it dynamic, you cannot just add a "if" statement in the mobiletable file. I need to check the viewport width whenever the size changes, meaning that I need to use the ResponsiveProvider code, and if I do that, I need to change everything else. Do you have any suggestions?

I can easily make this work for Iphone SE and below, but it will not be responsive.

MatissJanis commented on packages/desktop-client/src/components/budget/MobileBudgetTable.js:
```diff
@@ -37,6 +37,8 @@ import { AmountInput } from '../util/AmountInput';
 // import { DragDrop, Draggable, Droppable, DragDropHighlight } from './dragdrop';
 import { ListItem, ROW_HEIGHT } from './MobileTable';
 
+const smallVPWidth = document.documentElement.clientWidth < 365;
```
You can use `useResizeObserver` to listen for screen size changes and then show the blocks depending on that size.

Sample: https://github.com/actualbudget/actual/blob/21d5f117eec2e881c0dbc03ff85f02d63fccf3a8/packages/desktop-client/src/components/budget/MonthPicker.js#L43

MatissJanis replied:
I think this is my last comment. Once we have this solved I'll update the screenshots and I think we're good to go :)

Crazypkr1099 replied:
Unfortunately that will not work as its required to be in a function and not a class. As of right now, there is support for iphone SE and I truly don't believe people are going to resize their desktop screen less than 370px. If anyone wants to take over and add dynamic support, they can.

joel-jeremy commented:
This looks cool. Doesn't look as cramped on my phone but the category names have less space now. Maybe we should add text overflow handling.
![Screenshot_20230906-185355](https://github.com/actualbudget/actual/assets/20313680/4475fb4b-fc9f-4276-b2f4-54f77ab11a08)

Crazypkr1099 commented:
I agree Joel, but I feel this is a separate issue that we can work on later.

On Wed., Sep. 6, 2023, 9:59 p.m. Joel Jeremy Marquez, <
***@***.***> wrote:

> This looks cool. Doesn't look as cramped on my phone but the category
> names have less space now. Maybe we should add text overflow handling.
> [image: Screenshot_20230906-185355]
> <https://user-images.githubusercontent.com/20313680/266181219-4475fb4b-fc9f-4276-b2f4-54f77ab11a08.png>
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/actualbudget/actual/pull/1651#issuecomment-1709353756>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACYH3B4YTDDLEY5HJOCMEILXZETBRANCNFSM6AAAAAA4KGZKSE>
> .
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
>

MatissJanis commented on packages/desktop-client/src/components/budget/MobileBudgetTable.js:
```diff
@@ -774,10 +850,25 @@ class BudgetGroups extends Component {
+    window.addEventListener('resize', handleResize);
+
+    return () => {
+      window.removeEventListener('resize', handleResize);
```
<!-- Suggestions are specific requests to improve the subject under review. It is assumed that we all want to do what's best, so these comments are never dismissed as “mere suggestions”, but are taken seriously.  -->
**💬 suggestion:** ‏this is a great start. But could you move it to the highest level component in this file and then just pass down as a prop in order to reduce the duplicated logic?

Thanks

MatissJanis commented:
Updated your branch with `master`. This solves the VRT issues. :)

Crazypkr1099 commented:
Hey @MatissJanis , 

Is there anything else you would like me to do for this PR request. I have done the following:
- Added Spent column to mobile view
- Removed Spent column when less than 365 px for support on very small phones
- Added a toggle button to switch between spent/budgeted view for very small phones

Please let me know thank you.

trevdor commented on packages/desktop-client/src/components/budget/MobileBudgetTable.js:
```diff
@@ -416,16 +439,52 @@ class TotalsRow extends PureComponent {
-            }}
-            type="financial"
-          />
+          {!smallVPWidth ? (
```
Similar thoughts as in [the block of ternaries above](https://github.com/actualbudget/actual/pull/1651/files#diff-60e094a44d38fac03fb722c1cf8eb9ddafcf42fbd2d15c0c9f4c3b7d1ec90474R294). Can you consolidate the logic & remove duplication of components?

trevdor commented on packages/desktop-client/src/components/budget/MobileBudgetTable.js:
```diff
@@ -289,30 +284,59 @@ class BudgetCategory extends PureComponent {
-            month={month}
-            onBudgetAction={onBudgetAction}
-          />
+          {!smallVPWidth ? (
+            <BudgetCell
+              name="budgeted"
+              binding={budgeted}
+              editing={editing}
+              style={{ width: 90 }}
+              textStyle={{ ...styles.smallText, textAlign: 'right' }}
+              categoryId={category.id}
+              month={month}
+              onBudgetAction={onBudgetAction}
+            />
+          ) : !showBudgetView ? (
+            <BudgetCell
+              name="budgeted"
+              binding={budgeted}
+              editing={editing}
+              style={{ width: 90 }}
+              textStyle={{ ...styles.smallText, textAlign: 'right' }}
+              categoryId={category.id}
+              month={month}
+              onBudgetAction={onBudgetAction}
+            />
+          ) : null}
```
```suggestion
          {!smallVPWidth || !showBudgetView ? (
            <BudgetCell
              name="budgeted"
              binding={budgeted}
              editing={editing}
              style={{ width: 90 }}
              textStyle={{ ...styles.smallText, textAlign: 'right' }}
              categoryId={category.id}
              month={month}
              onBudgetAction={onBudgetAction}
            />
          ) : null}
```

trevdor commented on packages/desktop-client/src/components/budget/MobileBudgetTable.js:
```diff
@@ -815,11 +894,41 @@ export function BudgetTable(props) {
 
+  const toggleMobileDisplayPref = useSelector(state => {
+    return (
+      (state.prefs.local && state.prefs.local.toggleMobileDisplayPref) || true
```
The `MobileDisplay` in `toggleMobileDisplayPref` is pretty broad and I'm also not sure I follow the "toggle" part. 

If we're directly reading this pref to decide which column to show, perhaps `mobileShowBudgetedColPref`? Then the `useState` variable could match very closely -- `showBudgetedCol`, for instance. 

Alternatively, something like `mobileBudgetFirstColPref` set to a string value `budgeted` / `spent` might be extra easy to follow through the code. 🤔

trevdor commented on packages/desktop-client/src/components/budget/MobileBudgetTable.js:
```diff
@@ -815,11 +894,41 @@ export function BudgetTable(props) {
+    if (!showBudgetView) {
+      // remember the pref indefinitely
+      dispatch(savePrefs({ toggleMobileDisplayPref: true }));
+    } else {
+      // Set a cookie for 5 minutes
+      let d = new Date();
+      d.setTime(d.getTime() + 1000 * 60 * 5);
+      document.cookie =
+        'toggleMobileDisplayPref=true;path=/;expires=' + d.toUTCString();
+    }
```
I'd say it's safe to drop the `else` altogether and just indefinitely show the column the person picked last. I do think it's a nice touch to remember it for them. 👏🏻

trevdor commented on packages/desktop-client/src/components/budget/MobileBudgetTable.js:
```diff
@@ -847,23 +956,80 @@ export function BudgetTable(props) {
             />
           )}
           <View style={{ flex: 1 }} />
-
-          <View style={{ width: 90, justifyContent: 'center' }}>
-            <Label title="BUDGETED" style={{ color: colors.n1 }} />
-            <CellValue
-              binding={reportBudget.totalBudgetedExpense}
-              type="financial"
+          {!smallVPWidth ? (
```
Same deduplication request for the ternaries from here to ln 1032.

trevdor commented on packages/desktop-client/src/components/budget/MobileBudgetTable.js:
```diff
@@ -986,6 +1159,34 @@ function BudgetHeader({
         backgroundColor: colors.p5,
       }}
     >
+      {smallVPWidth ? (
```
Having this up in the header feels a little too cramped for my taste, especially for a screen that's already super narrow. (It's also on the side of the month heading further away from the column that it affects.)

Personally, my intuition is to tap the BUDGETED/SPENT header to change it. For discoverability, maybe add a subtle indicator that would at least suggest something can be done with the header. Something like this maybe? In the order `BUDGETED - SPENT - BALANCE`, the arrow points from BUDGETED toward where SPENT would be, and vice versa. (I haven't removed the header toggle text button here, but that would go.)

![budgeted-spent](https://github.com/actualbudget/actual/assets/5862724/a9ceb04c-90c7-4f3a-812b-b6db53d85348)


CSS for the div surrounding the BUDGETED/SPENT text is roughly:
```css
// padding + negative margin to keep the column text aligned
padding: '0 8px';
margin: '0 -8px'; 
background: `linear-gradient(-45deg, ${colors.p5} 8px, transparent 0)`;
// 45deg to flip it to the lower left corner
```

trevdor reviewed (changes requested):
Thanks for building this! I love the concept -- this is gonna be a cool feature.

But to land it in the app, we'll need to tweak the approach a bit. I tried to be detailed in my suggestions, but let me know if anything I've requested is unclear.

Crazypkr1099 commented on packages/desktop-client/src/components/budget/MobileBudgetTable.js:
```diff
@@ -788,11 +856,22 @@ class BudgetGroups extends Component {
+    return () => {
+      window.removeEventListener('resize', handleResize);
+    };
+  }, []);
```
I didn't know it was this easy to get the width boolean! Thank you for this!

trevdor commented on packages/desktop-client/src/components/budget/MobileBudgetTable.js:
```diff
@@ -815,11 +834,39 @@ export function BudgetTable(props) {
+  }
+  let buttonStyle = {
+    padding: 0,
+    backgroundColor: 'transparent',
```
The corner triangle gets a little cut off with the button radius.
```suggestion
    backgroundColor: 'transparent',
    borderRadius: 'unset',
```

trevdor reviewed (changes requested):
@Crazypkr1099 You took a load of feedback in stride. Impressively quick turnaround. I think it's looking great! Left one small change suggestion and then I'm good.

Crazypkr1099 commented:
> @Crazypkr1099 You took a load of feedback in stride. Impressively quick turnaround. I think it's looking great! Left one small change suggestion and then I'm good.

Great suggestion I totally missed that!

Thanks can't wait to see this added.

trevdor reviewed (approved):
Thanks again!

MatissJanis commented:
Sadly I can't push the new screenshot myself to your repo as I don't have the access. But here it is. Can you please upload it in place of `Mobile-loads-the-budget-page-with-budgeted-amounts-1-chromium-linux.png`?

![Mobile-loads-the-budget-page-with-budgeted-amounts-1-chromium-linux](https://github.com/actualbudget/actual/assets/886567/f1e55894-ce2f-4512-b307-253967bd4417)

MatissJanis reviewed (approved):
Thanks!
